amethyst_ui/src/pass.rs

```diff
@@ -0,0 +1,275 @@
+    type Apply = DrawUiApply<'a>;
+}
+
+impl Pass for DrawUi {
+    fn compile(&self, effect: NewEffect) -> Result<Effect> {
+        use std::mem;
+        effect
+            .simple(VERT_SRC, FRAG_SRC)
+            .with_raw_constant_buffer("VertexArgs", mem::size_of::<VertexArgs>(), 1)
+            .with_raw_vertex_buffer(PosTex::ATTRIBUTES, PosTex::size() as ElemStride, 0)
+            .with_texture("albedo")
+            .with_blended_output("color", ColorMask::all(), blend::ALPHA, None)
+            .build()
+    }
+
+    fn apply<'a, 'b: 'a>(
+        &'a mut self,
+        supplier: Supplier<'a>,
+        (entities, screen_dimensions, mesh_storage, tex_storage, ui_image, ui_transform, ui_text): (
+            Entities<'a>,
+            Fetch<'a, ScreenDimensions>,
+            Fetch<'a, AssetStorage<Mesh>>,
+            Fetch<'a, AssetStorage<Texture>>,
+            ReadStorage<'a, UiImage>,
+            ReadStorage<'a, UiTransform>,
+            ReadStorage<'a, UiText>,
+        ),
+) -> DrawUiApply<'a>{
+        DrawUiApply {
+            entities,
+            screen_dimensions,
+            mesh_storage,
+            tex_storage,
+            ui_image,
+            ui_transform,
+            ui_text,
+            unit_mesh: self.mesh_handle.clone(),
+            cached_draw_order: &mut self.cached_draw_order,
+            supplier,
+        }
+    }
+}
+
+pub struct DrawUiApply<'a> {
+    entities: Entities<'a>,
+    screen_dimensions: Fetch<'a, ScreenDimensions>,
+    mesh_storage: Fetch<'a, AssetStorage<Mesh>>,
+    tex_storage: Fetch<'a, AssetStorage<Texture>>,
+    ui_image: ReadStorage<'a, UiImage>,
+    ui_transform: ReadStorage<'a, UiTransform>,
+    ui_text: ReadStorage<'a, UiText>,
+    unit_mesh: MeshHandle,
+    cached_draw_order: &'a mut Vec<(f32, Entity)>,
+    supplier: Supplier<'a>,
+}
+
+impl<'a> ParallelIterator for DrawUiApply<'a> {
+    type Item = ();
+
+    fn drive_unindexed<C>(self, consumer: C) -> C::Result
+    where
+        C: UnindexedConsumer<Self::Item>,
+    {
+        let DrawUiApply {
+            entities,
+            screen_dimensions,
+            mesh_storage,
+            tex_storage,
+            ui_image,
+            ui_transform,
+            ui_text,
+            unit_mesh,
+            cached_draw_order,
+            supplier,
+            ..
+        } = self;
+
+        let entities = &*entities;
+        let screen_dimensions = &screen_dimensions;
+        let mesh_storage = &mesh_storage;
+        let tex_storage = &tex_storage;
+        let ui_image = &ui_image;
+        let ui_text = &ui_text;
+        let ui_transform = &ui_transform;
+        let unit_mesh = &unit_mesh;
+
+        // Populate and update the draw order cache.
+        // TODO: Replace all of this with code taking advantage of specs::TrackedStorage.
+        // TrackedStorage doesn't exist yet but it will in a later version of specs.
+        cached_draw_order.retain(|&(_z, entity)| {
```

Comment:
Is it really worth it to cache this data and not just recalculate it per frame ? The loop below is best case n^2, because position will potentially go through all items in the cache, plus a sort that's n*log(n). Wouldn't it be faster to use `binary_search_by` and do insertion sort and create the list from scratch every frame? That would be a guaranteed n*log(n).

Reply:
@Rhuagh Yeah you're right this eventually turned out to be not quite what was desired.  I have a couple ideas on how to fix it though.  First is to introduce a private AtomicBool to the UiTransform indicating if it was added to the cache or not, taking our n^2 operation down to just n.

Caching the sorting is valuable imo because sorting an already sorted collection is very cheap.

Reply:
I agree, it's just when the cache update is more expensive than rebuilding the whole thing I get suspicious :P TrackedStorage and insertion sort on z would turn it into nlogn worst case